### PR TITLE
issue 65 Add investments app

### PR DESCRIPTION
**Pull request by MaksymZymyn1989**
Implemented the Subscription model representing an investment by an investor into a project, with validation to prevent investing in own startup projects and unique constraints per investor-project pair.

Created the recalculate_investment_shares service function that recalculates and updates the investment_share percentage for all subscriptions of a given project, using efficient bulk update operations.

Added Django signals (post_save and post_delete) on Subscription to automatically trigger recalculation of investment shares whenever a subscription is created, updated, or deleted.

Running the test is successful:

### CODE REVIEW

File: core/settings.py [hunk 1]
```diff
@@ -300,3 +301,4 @@
 MAX_IMAGE_DIMENSIONS = (5000, 5000)
 ALLOWED_IMAGE_MODES = ["RGB", "RGBA", "L"]
 
+TEST_RUNNER = 'django.test.runner.DiscoverRunner'
```
Comment (mehalyna) [inline core/settings.py:304]: Explicitly setting the test runner is fine, but unless you have custom requirements, this is Django's default. Is this line necessary?

File: investments/admin.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+@admin.register(Subscription)
+class SubscriptionAdmin(admin.ModelAdmin):
+    """
+    Admin configuration for the Subscription model.
+
+    Makes the 'investment_share' field read-only to prevent manual edits,
+    and displays key fields in the admin list view.
+    """
+    readonly_fields = ('investment_share',)
```
Comment (mehalyna) [inline investments/admin.py:13]: 👍  Marking `investment_share` as read-only in the admin is a good practice since it's calculated automatically.

File: investments/admin.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    list_display = ('investor', 'project', 'amount', 'investment_share', 'created_at')
```
Comment (mehalyna) [inline investments/admin.py:14]: Consider adding `search_fields` or `list_filter` for more admin usability, especially if you expect many subscriptions.

File: investments/apps.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    such as signal registration when the app is ready.
+    """
+    name = 'investments'
+
+    def ready(self):
+        """
+        Called when the Django app registry is fully populated.
+        This method imports and registers signal handlers for the app.
+        """
+        import investments.signals  # noqa
```
Comment (mehalyna) [inline investments/apps.py:17]: _Just comment:_ Good use of `ready()` for signal registration. If this app gets larger, consider lazy-loading signals to avoid circular imports during migrations.

File: investments/models.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            this subscription represents in the project. Calculated automatically.
+        created_at (DateTime): Timestamp when the subscription was created.
+
+    Constraints:
+        - Investors cannot invest in their own startup's project.
+        - Each investor can have only one subscription per project (unique constraint).
+    """
+
+    investor = models.ForeignKey('profiles.Investor', on_delete=models.CASCADE)
+    project = models.ForeignKey('projects.Project', on_delete=models.CASCADE)
```
Comment (mehalyna) [inline investments/models.py:26]: Consider adding `related_name` attributes for easier reverse lookups.

File: investments/models.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    amount = models.DecimalField(
+        max_digits=18,
+        decimal_places=2,
+        validators=[MinValueValidator(0)],
```
Comment (mehalyna) [inline investments/models.py:30]: Should amounts of zero be allowed? If not, consider `MinValueValidator(Decimal("0.01"))` to enforce positive investments only.

File: investments/models.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    def clean(self):
+        """
+        Custom validation to prevent investors from investing in their own projects.
+
+        Raises:
+            ValidationError: If the investor is the owner of the project's startup.
+        """
+        if self.investor and self.project and self.project.startup.user_id == self.investor.user_id:
+            raise ValidationError("Investors cannot invest in their own startup's project.")
```
Comment (mehalyna) [inline investments/models.py:50]: 👍 Good business rule enforcement! 
Is this check robust to users that might not exist or be set? Consider handling possible `None` values for related objects.

File: investments/models.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    class Meta:
+        db_table = "subscriptions"
+        ordering = ["-created_at"]
+        verbose_name = "Subscription"
+        verbose_name_plural = "Subscriptions"
+        constraints = [
+            models.UniqueConstraint(
+                fields=["investor", "project"],
+                name="unique_investor_project"
+            ),
```
Comment (mehalyna) [inline investments/models.py:65]: 👍 Great use of unique constraints to prevent duplicate subscriptions.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: investments/apps.py (hunk 4 above)
Comment (MaksymZymyn1989) [reply investments/apps.py]: Thank you. I added lazy-loading with Celery  in the last commit.

File: investments/models.py (hunk 5 above)
Comment (MaksymZymyn1989) [reply investments/models.py]: Thank you. I added this in the last commit.

File: investments/models.py (hunk 6 above)
Comment (MaksymZymyn1989) [reply investments/models.py]: Thank you. I corrected this in the last commit.

File: investments/models.py (hunk 7 above)
Comment (MaksymZymyn1989) [reply investments/models.py]: Thank you. I added this in the last commit.

File: investments/services/investment_share_service.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
+from decimal import Decimal, ROUND_HALF_UP
+from django.db import models
+from investments.models import Subscription
+
+
+def recalculate_investment_shares(project):
```
[the remark this answers is not in the document]
Comment (MaksymZymyn1989) [reply investments/services/investment_share_service.py:5]: Thank you. I added doc-strings in the last commit.

File: investments/tests/test_services.py
[the remark this answers is not in the document]
Comment (MaksymZymyn1989) [reply investments/tests/test_services.py:1]: Thank you. I added this in the last commit.

File: investments/tests/test_services.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        s1 = Subscription.objects.create(project=self.project, investor=self.investor1, amount=Decimal('100'))
+        s2 = Subscription.objects.create(project=self.project, investor=self.investor2, amount=Decimal('300'))
+
+        recalculate_investment_shares(self.project)
+
+        s1.refresh_from_db()
+        s2.refresh_from_db()
+
+        self.assertEqual(s1.investment_share, Decimal('25.00'))
+        self.assertEqual(s2.investment_share, Decimal('75.00'))
```
[the remark this answers is not in the document]
Comment (MaksymZymyn1989) [reply investments/tests/test_services.py:106]: Thank you. I added this in the last commit.

**Review by Raidriar28 — APPROVED** [review]
I’ve thoroughly reviewed your implementation and honestly couldn’t find anything to improve or add.
Excellent implementation: the Subscription model is well-structured, with business validation that prevents investors from funding their own projects and a unique constraint per investor-project pair. The recalculate_investment_shares service is precise and reliable—using Decimal with ROUND_HALF_UP for financial accuracy, idempotent logic, and efficient bulk_update operations. Django signals (post_save and post_delete) are properly connected via apps.get_model, and the recalculation is triggered asynchronously through Celery, keeping request threads lightweight. The test suite is especially impressive—covering edge cases, rounding precision, zero-amount scenarios, and total share integrity.
Great job! The code is clean, well-structured, and covers all critical aspects—from validation and constraints to asynchronous recalculation and comprehensive testing.
